File: src/PythonClient/lib/quixstreaming/models/streamwriter/streameventswriter.py
from datetime import datetime, timedelta
from typing import Union
from quixstreaming.builders import *
from quixstreaming.models.netdict import NetDict
from quixstreaming import __importnet, EventData
from quixstreaming.helpers import *
import Quix.Sdk.Streaming
import pandas as pd

from quixstreaming.models.parametersbuffer import ParametersBuffer


class StreamEventsWriter(object):
    """
        Group all the Events properties, builders and helpers that allow to stream event values and event definitions to the platform.
    """

    def __init__(self, net_object: Quix.Sdk.Streaming.Models.StreamWriter.StreamEventsWriter = None):
        """
            Initializes a new instance of StreamEventsWriter.

            Events:

            net_object (.net object): The .net object representing StreamEventsWriter.
        """

        if net_object is None:
            raise Exception("StreamEventsWriter is none")
        else:
            self.__wrapped = net_object

    def flush(self):
        """
        Immediately writes the event definitions from the buffer without waiting for buffer condition to fulfill (200ms timeout)
        """
        self.__wrapped.Flush()

    @property
    def default_tags(self) -> NetDict:
        """Get default tags injected to all Events Values sent by the writer."""
        return NetDict(self.__wrapped.DefaultTags)

    @property
    def default_location(self) -> str:
        """
            Gets the default Location of the events. Event definitions added with add_definition  will be inserted at this location.
            See add_location for adding definitions at a different location without changing default.
            Example: "/Group1/SubGroup2"
        """
        return self.__wrapped.DefaultLocation

    @default_location.setter
    def default_location(self, value: str):
        """
            Sets the default Location of the events. Event definitions added with add_definition will be inserted at this location.
            See add_location for adding definitions at a different location without changing default.
            Example: "/Group1/SubGroup2"
        """
        self.__wrapped.DefaultLocation = value

    @property
    def epoch(self) -> datetime:
        """Get the default epoch used for event values"""
        return NetToPythonConverter.convert_datetime(self.__wrapped.Epoch)

    @epoch.setter
    def epoch(self, value: datetime):
        """Set the default epoch used for event values"""
        self.__wrapped.Epoch = PythonToNetConverter.convert_datetime(value)

    def write(self, data: Union[EventData, pd.DataFrame], **columns) -> None:
        """Writes event into the stream.

        Parameters: data: EventData object or a Pandas dataframe. columns: Column names if the dataframe has
        different columns from 'id', 'timestamp' and 'value'. For instance if 'id' is in the column 'event_id',
        id='event_id' must be passed as an argument.

        Raises:
            TypeError if the data argument is neither an EventData nor Pandas dataframe.
        """
        if isinstance(data, EventData):
            net_object = data.convert_to_net()
            self.__wrapped.Write(net_object)
        elif isinstance(data, pd.DataFrame):
            id = 'id' if 'id' not in columns else columns['id']
            timestamp = 'timestamp' if 'timestamp' not in columns else columns['timestamp']
            value = 'value' if 'value' not in columns else columns['value']
            for row in data.itertuples():
                event = EventData(event_id=getattr(row, id), time=getattr(row, timestamp), value=getattr(row, value))
                obj = event.convert_to_net()
                self.__wrapped.Write(obj)
        else:
            raise TypeError(str(type(data)) + " is not supported.")


    def add_timestamp(self, time: Union[datetime, timedelta]) -> EventDataBuilder:
        """
        Start adding a new set of event values at the given timestamp.
        :param time: The time to use for adding new event values.
                     | datetime: The datetime to use for adding new event values. NOTE, epoch is not used
                     | timedelta: The time since the default epoch to add the event values at

        :return: EventDataBuilder
        """
        if time is None:
            raise ValueError("'time' must not be None")
        if isinstance(time, datetime):
            netdate = PythonToNetConverter.convert_datetime(time)
            return EventDataBuilder(self.__wrapped.AddTimestamp(netdate))
        if isinstance(time, timedelta):
            nettimespan = PythonToNetConverter.convert_timedelta(time)
            return EventDataBuilder(self.__wrapped.AddTimestamp(nettimespan))
        raise ValueError("'time' must be either datetime or timedelta")

    def add_timestamp_milliseconds(self, milliseconds: int) -> EventDataBuilder:
        """
        Start adding a new set of event values at the given timestamp.
        :param milliseconds: The time in milliseconds since the default epoch to add the event values at
        :return: EventDataBuilder
        """
        return EventDataBuilder(self.__wrapped.AddTimestampMilliseconds(milliseconds))

    def add_timestamp_nanoseconds(self, nanoseconds: int) -> EventDataBuilder:
        """
        Start adding a new set of event values at the given timestamp.
        :param nanoseconds: The time in nanoseconds since the default epoch to add the event values at
        :return: EventDataBuilder
        """
        return EventDataBuilder(self.__wrapped.AddTimestampNanoseconds(nanoseconds))

    def add_definition(self, event_id: str, name: str = None, description: str = None) -> EventDefinitionBuilder:
        """
        Add new event definition to the StreamPropertiesWriter. Configure it with the builder methods.
        :param event_id: The id of the event. Must match the event id used to send data.
        :param name: The human friendly display name of the event
        :param description: The description of the event
        :return: EventDefinitionBuilder to define properties of the event or add additional events
        """
        return EventDefinitionBuilder(self.__wrapped.AddDefinition(event_id, name, description))

    def add_location(self, location: str) -> EventDefinitionBuilder:
        """
        Add a new location in the events groups hierarchy
        :param location: The group location
        :return: EventDefinitionBuilder to define the events under the specified location
        """
        return EventDefinitionBuilder(self.__wrapped.AddLocation(location))

File: src/CsharpClient/Quix.Sdk.Streaming/Models/StreamWriter/StreamEventsWriter.cs
using System;
using System.Collections.Generic;
using System.Threading;
using Microsoft.Extensions.Logging;
using Quix.Sdk.Process.Models;
using Quix.Sdk.Process.Models.Utility;
using Quix.Sdk.Process.Managers;
using System.Linq;
using Quix.Sdk.Streaming.Exceptions;

namespace Quix.Sdk.Streaming.Models.StreamWriter
{
    /// <summary>
    /// Helper class for writing <see cref="EventDefinitions"/> and <see cref="EventData"/>
    /// </summary>
    public class StreamEventsWriter : IDisposable
    {
        private readonly ILogger logger = Logging.CreateLogger<StreamEventsWriter>();
        private readonly IStreamWriterInternal streamWriter;

        private long epoch = 0;

        private string location;
        private readonly EventDefinitionsManager eventDefinitionsManager = new EventDefinitionsManager();
        private readonly Timer flushDefinitionsTimer;
        private bool timerEnabled = false; // Here because every now and then reseting its due time to never doesn't work
        private const int TimerInterval = 200;
        private readonly object flushLock = new object();
        private bool isDisposed;

        /// <summary>
        /// Initializes a new instance of <see cref="StreamEventsWriter"/>
        /// </summary>
        /// <param name="streamWriter">Stream writer owner</param>
        internal StreamEventsWriter(IStreamWriterInternal streamWriter)
        {
            this.streamWriter = streamWriter;

            // Timer for Flush Parameter definitions
            flushDefinitionsTimer = new Timer(OnFlushDefinitionsTimerEvent, null, Timeout.Infinite, Timeout.Infinite);

            // Initialize root location
            this.DefaultLocation = "/";
        }

        /// <summary>
        /// Default Tags injected to all Event Values sent by the writer.
        /// </summary>
        public Dictionary<string, string> DefaultTags { get; set; } = new Dictionary<string, string>();

        /// <summary>
        /// Default Location of the events. Event definitions added with <see cref="AddDefinition"/> will be inserted at this location.
        /// See <see cref="AddLocation"/> for adding definitions at a different location without changing default.
        /// Example: "/Group1/SubGroup2"
        /// </summary>
        public string DefaultLocation
        {
            get
            {
                return this.location;
            }
            set
            {
                if (isDisposed)
                {
                    throw new ObjectDisposedException(nameof(StreamEventsWriter));
                }
                this.location = this.eventDefinitionsManager.ReformatLocation(value);
            }
        }

        /// <summary>
        /// Default Epoch used for Timestamp event values. Datetime added on top of all the Timestamps.
        /// </summary>
        public DateTime Epoch
        {
            get
            {
                return epoch.FromUnixNanoseconds();
            }
            set
            {
                if (isDisposed)
                {
                    throw new ObjectDisposedException(nameof(StreamEventsWriter));
                }
                epoch = value.ToUnixNanoseconds();
            }
        }

        /// <summary>
        /// Starts adding a new set of event values at the given timestamp.
        /// Note, <see cref="Epoch"/> is not used when invoking with <see cref="DateTime"/>
        /// </summary>
        /// <param name="dateTime">The datetime to use for adding new event values</param>
        /// <returns>Event data builder to add event values at the provided time</returns>
        public EventDataBuilder AddTimestamp(DateTime dateTime) => this.AddTimestampNanoseconds(dateTime.ToUnixNanoseconds(), 0);

        /// <summary>
        /// Starts adding a new set of event values at the given timestamp.
        /// </summary>
        /// <param name="timeSpan">The time since the default <see cref="Epoch"/> to add the event values at</param>
        /// <returns>Event data builder to add event values at the provided time</returns>
        public EventDataBuilder AddTimestamp(TimeSpan timeSpan) => this.AddTimestampNanoseconds(timeSpan.ToNanoseconds());

        /// <summary>
        /// Starts adding a new set of event values at the given timestamp.
        /// </summary>
        /// <param name="timeMilliseconds">The time in milliseconds since the default <see cref="Epoch"/> to add the event values at</param>
        /// <returns>Event data builder to add event values at the provided time</returns>
        public EventDataBuilder AddTimestampMilliseconds(long timeMilliseconds) => this.AddTimestampNanoseconds(timeMilliseconds * (long) 1e6);

        /// <summary>
        /// Starts adding a new set of event values at the given timestamp.
        /// </summary>
        /// <param name="timeNanoseconds">The time in nanoseconds since the default <see cref="Epoch"/> to add the event values at</param>
        /// <returns>Event data builder to add event values at the provided time</returns>
        public EventDataBuilder AddTimestampNanoseconds(long timeNanoseconds)
        {
            if (isDisposed)
            {
                throw new ObjectDisposedException(nameof(StreamEventsWriter));
            }
            return AddTimestampNanoseconds(timeNanoseconds, this.epoch);
        }
        
        private EventDataBuilder AddTimestampNanoseconds(long timeNanoseconds, long epoch)
        {
            if (isDisposed)
            {
                throw new ObjectDisposedException(nameof(StreamEventsWriter));
            }
            return new EventDataBuilder(this, epoch + timeNanoseconds);
        }

        /// <summary>
        /// Adds a list of definitions to the <see cref="StreamEventsWriter"/>. Configure it with the builder methods.
        /// </summary>
        /// <param name="definitions">List of definitions</param>
        public void AddDefinitions(List<EventDefinition> definitions)
        {
            if (isDisposed)
            {
                throw new ObjectDisposedException(nameof(StreamEventsWriter));
            }
            definitions.ForEach(d => this.eventDefinitionsManager.AddDefinition(d.ConvertToProcessDefinition(), d.Location));

            this.ResetFlushDefinitionsTimer();
        }

        /// <summary>
        /// Adds new Event definition, to define properties like Name or Level, among others.
        /// </summary>
        /// <param name="eventId">Event Id. This must match the event id you use to Event values</param>
        /// <param name="name">Human friendly display name of the event</param>
        /// <param name="description">Description of the event</param>
        /// <returns>Event definition builder to define the event properties</returns>
        public EventDefinitionBuilder AddDefinition(string eventId, string name = null, string description = null)
        {
            if (isDisposed)
            {
                throw new ObjectDisposedException(nameof(StreamEventsWriter));
            }
            var eventDefinition = this.CreateDefinition(this.location, eventId, name, description);

            var builder = new EventDefinitionBuilder(this, this.location, eventDefinition);

            return builder;
        }

        /// <summary>
        /// Adds a new Location in the event groups hierarchy.
        /// </summary>
        /// <param name="location"></param>
        public EventDefinitionBuilder AddLocation(string location)
        {
            if (isDisposed)
            {
                throw new ObjectDisposedException(nameof(StreamEventsWriter));
            }
            this.eventDefinitionsManager.GenerateLocations(location);

            var builder = new EventDefinitionBuilder(this, location);

            return builder;
        }

        internal Process.Models.EventDefinition CreateDefinition(string location, string eventId, string name, string description)
        {
            if (isDisposed)
            {
                throw new ObjectDisposedException(nameof(StreamEventsWriter));
            }
            var eventDefinition = new Process.Models.EventDefinition
            {
                Id = eventId,
                Name = name,
                Description = description
            };

            eventDefinitionsManager.AddDefinition(eventDefinition, location);

            this.ResetFlushDefinitionsTimer();

            return eventDefinition;
        }


        /// <summary>
        /// Immediately writes the event definitions from the buffer without waiting for buffer condition to fulfill (200ms timeout)
        /// </summary>
        public void Flush()
        {
            this.Flush(false);
        }

        private void Flush(bool force)
        {
            if (!force && isDisposed)
            {
                throw new ObjectDisposedException(nameof(StreamEventsWriter));
            }
            try
            {
                lock (flushLock)
                {
                    this.FlushDefinitions();
                }
            }
            catch (Exception ex)
            {
                this.logger.LogError(ex, "Exception occurred while trying to flush events data buffer.");
            }
        }

        /// <summary>
        /// Write event into the stream
        /// </summary>
        /// <param name="data">Event data to write</param>
        public void Write(EventData data)
        {
            if (isDisposed)
            {
                throw new ObjectDisposedException(nameof(StreamEventsWriter));
            }
            if (!data.EpochIncluded)
            {
                data.TimestampNanoseconds += this.Epoch.ToUnixNanoseconds();
                data.EpochIncluded = true;
            }

            foreach (var kv in this.DefaultTags)
            {
                if (!data.Tags.ContainsKey(kv.Key))
                {
                    data.AddTag(kv.Key, kv.Value);
                }
            }

            this.streamWriter.Write(data.ConvertToProcessData());
            this.logger.Log(LogLevel.Trace, "event '{0}' sent.", data.Id);
        }

        /// <summary>
        /// Write events into the stream
        /// </summary>
        /// <param name="events">Events to write</param>
        public void Write(ICollection<EventData> events)
        {
            if (isDisposed)
            {
                throw new ObjectDisposedException(nameof(StreamEventsWriter));
            }
            foreach(var data in events)
            {
                foreach (var kv in this.DefaultTags)
                {
                    if (!data.Tags.ContainsKey(kv.Key))
                    {
                        data.AddTag(kv.Key, kv.Value);
                    }
                }
            }

            var batch = events.Select(e => e.ConvertToProcessData()).ToArray();

            this.streamWriter.Write(batch);
            this.logger.Log(LogLevel.Trace, "{0} event(s) sent.", events.Count);
        }

        internal void Write(ICollection<Process.Models.EventDataRaw> events)
        {
            if (isDisposed)
            {
                throw new ObjectDisposedException(nameof(StreamEventsWriter));
            }
            this.streamWriter.Write(events);
        }

        private void ResetFlushDefinitionsTimer()
        {
            if (isDisposed) return;
            timerEnabled = true;
            flushDefinitionsTimer.Change(TimerInterval, Timeout.Infinite);
        }

        private void OnFlushDefinitionsTimerEvent(object state)
        {
            if (!timerEnabled) return;
            try
            {
                this.FlushDefinitions();
            }
            catch (StreamClosedException exception) when (this.isDisposed)
            {
                // Ignore exception because the timer flush definition may finish executing only after closure due to how close lock works in streamWriter
            }
            catch (Exception ex)
            {
                this.logger.Log(LogLevel.Error, ex, "Exception occurred while trying to flush event definition buffer.");
            }
        }


        private void FlushDefinitions()
        {
            timerEnabled = false;
            flushDefinitionsTimer.Change(Timeout.Infinite, Timeout.Infinite);

            var definitions = eventDefinitionsManager.GenerateEventDefinitions();
            
            if (definitions.Events?.Count == 0 && definitions.EventGroups?.Count == 0) return; // there is nothing to flush

            this.streamWriter.Write(definitions);
        }

        /// <summary>
        /// Flushes internal buffers and disposes
        /// </summary>
        public void Dispose()
        {
            if (this.isDisposed) return;
            this.isDisposed = true;
            this.Flush(true);
            flushDefinitionsTimer?.Dispose();
        }
    }
}
